File: .reasonix/skills/abacatepay/examples/ts/mrr.ts
import { AbacatePay } from '@abacatepay/sdk';

// Define types for MRR operations
export type MrrData = {
  mrr: number;
  totalActiveSubscriptions: number;
};

export type MerchantInfo = {
  name: string;
  website: string;
  createdAt: string;
};

export type Transaction = {
  amount: number;
  count: number;
};

export type RevenueData = {
  totalRevenue: number;
  totalTransactions: number;
  transactionsPerDay: Record<string, Transaction>;
  revenue: number;
  currency: string;
};

const abacate = AbacatePay({ secret: process.env.ABACATEPAY_API_KEY });

/**
 * Retrieves Monthly Recurring Revenue (MRR) data.
 * Shows recurring revenue metrics for business insights.
 * @returns Promise<MRRData[]> - Array of MRR data points
 */
export async function getMRR(): Promise<MRRData[]> {
  try {
    // Fetch MRR metrics using SDK
    const data = await abacate.mrr.get();

    console.log("MRR Data retrieved:", data.length, "entries");
    return data;
  } catch (error) {
    console.error('Error fetching MRR:', error);
    throw error;
  }
}

/**
 * Retrieves merchant account information.
 * Useful for account verification and profile data.
 * @returns Promise<MerchantInfo> - Merchant account details
 */
export async function getMerchantInfo(): Promise<MerchantInfo> {
  try {
    // Fetch merchant account information using SDK
    const data = await abacate.mrr.merchantInfo();

    console.log("Merchant Info retrieved:", data.name);
    return data;
  } catch (error) {
    console.error('Error fetching Merchant Info:', error);
    throw error;
  }
}

/**
 * Retrieves revenue data and analytics.
 * Provides total revenue metrics for business reporting.
 * @returns Promise<RevenueData[]> - Array of revenue data points
 */
export async function getRevenue(): Promise<RevenueData[]> {
  try {
    // Fetch revenue analytics using SDK
    const data = await abacate.mrr.revenue();

    console.log("Revenue Data retrieved:", data.length, "entries");
    return data;
  } catch (error) {
    console.error('Error fetching Revenue:', error);
    throw error;
  }
}

File: .reasonix/skills/abacatepay/examples/go/mrr.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// Define structs for MRR operations
type MrrData struct {
	MRR                      float64 `json:"mrr"`
	TotalActiveSubscriptions int     `json:"totalActiveSubscriptions"`
}

type MrrResponse struct {
	Data    *MrrData `json:"data,omitempty"`
	Error   *string  `json:"error"`
	Success bool     `json:"success"`
}

type MerchantInfo struct {
	Name      string `json:"name"`
	Website   string `json:"website"`
	CreatedAt string `json:"createdAt"`
}

type Transaction struct {
	Amount int `json:"amount"`
	Count  int `json:"count"`
}

type RevenueData struct {
	TotalRevenue       int                    `json:"totalRevenue"`
	TotalTransactions  int                    `json:"totalTransactions"`
	TransactionsPerDay map[string]Transaction `json:"transactionsPerDay"`
	Revenue            float64                `json:"revenue"`
	Currency           string                 `json:"currency"`
}

var MRRBaseURL = BaseURL + "/public-mrr"

// Retrieves Monthly Recurring Revenue (MRR) data.
// Shows recurring revenue metrics for business insights.
// Returns array of MRR data points.
func GetMRR() (*DefaultResponse[[]MrrData], error) {
	apiKey := os.Getenv("ABACATEPAY_API_KEY")
	if apiKey == "" {
		return nil, fmt.Errorf("ABACATEPAY_API_KEY not set")
	}

	req, err := http.NewRequest("GET", MRRBaseURL+"/mrr", nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("api error: status %d - %s", resp.StatusCode, string(body))
	}

	var result DefaultResponse[[]MrrData]
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("error decoding response: %w", err)
	}

	return &result, nil
}

// Retrieves merchant account information.
// Useful for account verification and profile data.
// Returns merchant account details.
func GetMerchantInfo() (*DefaultResponse[MerchantInfo], error) {
	apiKey := os.Getenv("ABACATEPAY_API_KEY")
	if apiKey == "" {
		return nil, fmt.Errorf("ABACATEPAY_API_KEY not set")
	}

	req, err := http.NewRequest("GET", MRRBaseURL+"/merchant-info", nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("api error: status %d - %s", resp.StatusCode, string(body))
	}

	var result DefaultResponse[MerchantInfo]
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("error decoding response: %w", err)
	}

	return &result, nil
}

// Retrieves revenue data and analytics.
// Provides total revenue metrics for business reporting.
// Returns array of revenue data points.
func GetRevenue() (*DefaultResponse[[]RevenueData], error) {
	apiKey := os.Getenv("ABACATEPAY_API_KEY")
	if apiKey == "" {
		return nil, fmt.Errorf("ABACATEPAY_API_KEY not set")
	}

	req, err := http.NewRequest("GET", MRRBaseURL+"/revenue", nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error making request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("api error: status %d - %s", resp.StatusCode, string(body))
	}

	var result DefaultResponse[[]RevenueData]
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("error decoding response: %w", err)
	}

	return &result, nil
}
